src/rust/engine/process_execution/bazel_protos/src/lib.rs

```diff
@@ -9,3 +9,385 @@ pub mod operations_grpc;
 pub mod remote_execution;
 pub mod remote_execution_grpc;
 pub mod status;
+
+use std::collections::HashSet;
+
+pub fn verify_directory_canonical(directory: &remote_execution::Directory) -> Result<(), String> {
```

Comment:
Should these functions go in a `test_util` package? Or are we actually planning to use them at runtime?

Reply:
Actually going to be used - my intent is to refuse to put non-canonical Directories in the Store, to help guarantee that our cache keys are correct.